strawberryfields/backends/tfbackend/ops.py

```diff
@@ -307,92 +317,96 @@ def loss_superop(T, D, batched=False):
     return output
 
 
-def displacement_matrix(alpha, D, batched=False):
-    """creates the single mode displacement matrix"""
-    if batched:
-        batch_size = alpha.shape[0]
-    alpha = tf.cast(alpha, def_type)
-    idxs = [(j, k) for j in range(D) for k in range(j)]
-    values = [
-        alpha ** (j - k) * tf.cast(tf.sqrt(binom(j, k) / factorial(j - k)), def_type)
-        for j in range(D)
-        for k in range(j)
-    ]
-    values = tf.stack(values, axis=-1)
-    dense_shape = [D, D]
-    vals = [1.0] * D
-    ind = [(idx, idx) for idx in range(D)]
-    if batched:
-        dense_shape = [batch_size] + dense_shape
-        vals = vals * batch_size
-        ind = batchify_indices(ind, batch_size)
-    eye_diag = tf.SparseTensor(ind, vals, dense_shape)
-    signs = [(-1) ** (j - k) for j in range(D) for k in range(j)]
+@tf.custom_gradient
+def single_displacement_matrix(r, phi, D, dtype=def_type.as_numpy_dtype):
+    """creates a single mode displacement matrix"""
+    r = r.numpy()
+    phi = phi.numpy()
+    gate = displacement_tw(r, phi, D, dtype)
+
+    def grad(dy):
+        Dr, Dphi = grad_displacement_tw(gate, r, phi)
+        grad_r = tf.math.real(tf.reduce_sum(dy * tf.math.conj(Dr)))
+        grad_phi = tf.math.real(tf.reduce_sum(dy * tf.math.conj(Dphi)))
+        return grad_r, grad_phi, None
+
+    return gate, grad
+
+
+def displacement_matrix(r, phi, D, batched=False):
+    """creates a single mode displacement matrix accounting for batching"""
+    r = tf.cast(r, def_type)
+    phi = tf.cast(phi, def_type)
     if batched:
-        idxs = batchify_indices(idxs, batch_size)
-        signs = signs * batch_size
-    sign_lower_diag = tf.cast(tf.SparseTensor(idxs, signs, dense_shape), tf.float32)
-    sign_matrix = tf.sparse.add(eye_diag, sign_lower_diag)
-    sign_matrix = tf.cast(tf.sparse.to_dense(sign_matrix), def_type)
-    lower_diag = tf.scatter_nd(idxs, tf.reshape(values, [-1]), dense_shape)
-    E = tf.cast(tf.eye(D), def_type) + lower_diag
-    E_prime = tf.math.conj(E) * sign_matrix
+        return tf.stack(
+            [single_displacement_matrix(r_, phi_, D) for r_, phi_ in tf.transpose([r, phi])]
+        )
+    return single_displacement_matrix(r, phi, D)
+
+
+@tf.custom_gradient
+def single_beamsplitter_matrix(theta, phi, D, dtype=def_type.as_numpy_dtype):
+    """creates a single mode beamsplitter matrix"""
+    theta = theta.numpy()
+    phi = phi.numpy()
+
+    gate = beamsplitter_tw(theta, phi, D, dtype)
+    gate = np.transpose(gate, [0, 2, 1, 3])
+
+    def grad(dy):
+        Dtheta, Dphi = grad_beamsplitter_tw(gate, theta, phi)
+        Dtheta = np.transpose(Dtheta, [0, 2, 1, 3])
+        Dphi = np.transpose(Dphi, [0, 2, 1, 3])
+        grad_theta = tf.math.real(tf.reduce_sum(dy * tf.math.conj(Dtheta)))
+        grad_phi = tf.math.real(tf.reduce_sum(dy * tf.math.conj(Dphi)))
+        return grad_theta, grad_phi, None
```

Comment:
Hi @ziofil, it seems that this will need a test as it's not covered by the coverage

Reply:
So, the gradient function `grad_beamsplitter_tw` is tested in thewalrus, and we are testing gates and gradients here as well.
If you mean the inner def grad(dy) function, then I'm not sure how to test it since it's inside the decorator (`single_beamsplitter_matrix` does not actually return a tuple since it's decorated).

Reply:
Agreed, thanks for the details on this! :)

Reply:
I had a look at the integration tests and we are testing gradients of states related to the displacement and squeezing ops, but not the beamsplitter (which is tested in thewalrus though). If we want to add more gradient tests, perhaps a test of the beamsplitter gradients would be a good idea.

Reply:
btw, I'm working on this test.